### PR TITLE
Bookmark command allows other users to bookmark the same message

### LINKED ISSUE #341: Allow Users to React to a Bookmark to Recieve a Copy
Spun off of conversations in https://github.com/python-discord/seasonalbot/issues/338#issuecomment-578322144 and https://github.com/python-discord/seasonalbot/pull/340#pullrequestreview-348784550:

@kwzrd:
> I've also seen a suggestion that the bot could listen for further clicks on its 📨 reaction, and dispatch the bookmark to those who click as well. This would be a super clean way to ask to be given the bm as well, although it may not be very obvious that it does that.

@AG4lyf made a good initial implementation in #340 that exposed some UX concerns:

@sco1:
> Regarding the reaction by others to receive the same bookmark: the implementation details need to be discussed more before the feature is merged.
> 
> I don't think it makes sense from a UX perspective to utilize the same reaction for multiple purposes when one of those purposes silently disappears after a timeout. This is already barely discoverable functionality, and with this implementation a user that does know of its existence doesn't know if it has timed out or if it's broken. A second emoji solves half of the problem, which might be sufficient.

Since this implementation discussion shouldn't block the fix being implemented by #340, it has been removed from the scope of that PR.

Per discussions in Discord, the initial feeling is to provide a separate reaction to expose this functionality & keep it distinct from the bot's feedback that the DM has been sent. :pushpin: or :round_pushpin: have been suggested.

Also to be discussed is whether or not this should be implemented as either:
* A `wait_for` reaction loop, with a timeout that removes the reaction
* An `on_reaction_add` event handler with a persistent cache that links the bookmark invocation message and the target message.

The former is simpler to implement & is what I'm leaning towards. The latter is more generic but will be trickier to implement in a way that prevents the file from growing too large & limit vectors for trolling.

The reaction emoji and reaction handling approach should be decided on prior to work being done.

### PR DESCRIPTION
**Pull request by ChrisLovering**
## Relevant Issues

Closes #341

## Description

Removed the previous ✉ emoji that got added the to author's message.
Replaced it with the below functionality:
When someone calls `.bm` on a message the bot will:
- Send a DM as normal (with the same error embed if DMs are disabled)
- Send an embed explaining that you can react to the embed and also get sent a DM with a link
- Add a 📌 emoji to the embed
- Have a 2 min `wait_for` listening for 📌 reactions
- On reaction, the bot will send the bookmark DM to the reactor
- If the reactor has DMs closed, send the same error embed as above.
- Add the user to a list, so they can't remove & add reactions to spam the bot
- after the 2 minutes are up, delete the embed to save screen space. 

## Reasoning

The previous emoji was removed, as it was no longer needed. The new embed was added to ease discoverability of the new feature. 

## Screenshots

image

## Did you:

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] If dependencies have been added or updated, run `pipenv lock`?
- [x] **Lint your code** (`pipenv run lint`)?
- [x] Set the PR to **allow edits from contributors**?

### CODE REVIEW

**Review by Kronifer — APPROVED** [review]
Tested on my end, works great! Great job!

File: bot/exts/evergreen/bookmark.py [hunk 1]
```diff
@@ -1,21 +1,83 @@
...
 
 class Bookmark(commands.Cog):
     """Creates personal bookmarks by relaying a message link to the user's DMs."""
 
     def __init__(self, bot: commands.Bot):
         self.bot = bot
 
+    @staticmethod
+    def build_bookmark_dm(target_message: WrappedMessageConverter, title: str) -> discord.Embed:
+        """Builds the embed to DM the bookmark requestor."""
```
Comment (Shivansh-007) [inline bot/exts/evergreen/bookmark.py:25]:
```suggestion
        """Builds the embed to DM the bookmark requester."""
```

File: bot/exts/evergreen/bookmark.py [hunk 2]
```diff
@@ -1,21 +1,83 @@
...
+        )
+        embed.add_field(name="Wanna give it a visit?", value=f"[Visit original message]({target_message.jump_url})")
+        embed.set_author(name=target_message.author, icon_url=target_message.author.avatar_url)
+        embed.set_thumbnail(url=Icons.bookmark)
+
+        return embed
+
+    @staticmethod
+    def build_error_embed(user: discord.Member) -> discord.Embed:
+        """Builds an error embed for when a bookmark requestor has DMs disabled."""
```
Comment (Shivansh-007) [inline bot/exts/evergreen/bookmark.py:39]:
```suggestion
        """Builds an error embed for when a bookmark requester has DMs disabled."""
```

File: bot/exts/evergreen/bookmark.py [hunk 3]
```diff
@@ -37,27 +99,39 @@ async def bookmark(
...
+        bookmarked_users = [ctx.author.id]
+        reaction_message = await self.send_reaction_embed(ctx.channel, target_message)
+
+        while True:
+            try:
+                _, user = await self.bot.wait_for("reaction_add", timeout=TIMEOUT, check=event_check)
+            except asyncio.TimeoutError:
+                log.debug("Timed out waiting for a reaction")
+                break  # No reactions for the last `TIMEOUT` seconds
+            log.trace(f"{user} has sucessfully bookmarked from a reaction, attempting to DM them.")
```
Comment (Shivansh-007) [inline bot/exts/evergreen/bookmark.py:129]:
```suggestion
            log.trace(f"{user} has successfully bookmarked from a reaction, attempting to DM them.")
```

File: bot/exts/evergreen/bookmark.py [hunk 4]
```diff
@@ -1,21 +1,83 @@
...
 
+    @staticmethod
+    def build_bookmark_dm(target_message: WrappedMessageConverter, title: str) -> discord.Embed:
+        """Builds the embed to DM the bookmark requestor."""
+        embed = discord.Embed(
+            title=title,
+            description=target_message.content,
+            colour=Colours.soft_green
+        )
+        embed.add_field(name="Wanna give it a visit?", value=f"[Visit original message]({target_message.jump_url})")
```
Comment (Shivansh-007) [inline bot/exts/evergreen/bookmark.py:31]:
```suggestion
        embed.add_field(
            name="Wanna give it a visit?", 
            value=f"[Visit original message]({target_message.jump_url})"
        )
```

**Review by Shivansh-007 — CHANGES_REQUESTED** [review]

File: bot/exts/evergreen/bookmark.py [hunk 5]
```diff
@@ -37,27 +104,39 @@ async def bookmark(
...
-        embed.set_thumbnail(url=Icons.bookmark)
-
-        try:
-            await ctx.author.send(embed=embed)
-        except discord.Forbidden:
-            error_embed = discord.Embed(
-                title=random.choice(ERROR_REPLIES),
-                description=f"{ctx.author.mention}, please enable your DMs to receive the bookmark",
-                colour=Colours.soft_red
+        def event_check(reaction_: discord.Reaction, user_: discord.Member) -> bool:
```
Comment (Xithrius) [inline bot/exts/evergreen/bookmark.py:107]: Why are there underscores at the end of these variable names? They don't overwrite anything as far as I can see.

File: bot/exts/evergreen/bookmark.py [hunk 6]
```diff
@@ -37,27 +104,39 @@ async def bookmark(
...
+        # Keep track of who has already bookmarked, so users can't spam reactions and cause loads of DMs
+        bookmarked_users = [ctx.author.id]
+        reaction_message = await self.send_reaction_embed(ctx.channel, target_message)
+
+        while True:
+            try:
+                _, user = await self.bot.wait_for("reaction_add", timeout=TIMEOUT, check=event_check)
+            except asyncio.TimeoutError:
+                log.debug("Timed out waiting for a reaction")
+                break  # No reactions for the last `TIMEOUT` seconds
```
Comment (Xithrius) [inline bot/exts/evergreen/bookmark.py:133]: Just to be consistent with the rest of this file.
```suggestion
                log.debug("Timed out waiting for a reaction")
                # No reactions for the last `TIMEOUT` seconds
                break
```

**Review by Xithrius — COMMENTED** [review]
Small review for now, will do another after testing.

File: bot/exts/evergreen/bookmark.py (hunk 5 above)
Comment (ChrisLovering) [reply bot/exts/evergreen/bookmark.py:107]: copy&paste strikes again D:

**Review by Shivansh-007 — APPROVED** [review]
:shipit:

**Review by ToxicKidz — APPROVED** [review]
LGTM, one comment.

File: bot/exts/evergreen/bookmark.py [hunk 7]
```diff
@@ -1,21 +1,88 @@
...
 
+# Number of seconds to wait for other users to bookmark the same message
+TIMEOUT = 120
+BOOKMARK_EMOJI = "📌"
+
 
 class Bookmark(commands.Cog):
     """Creates personal bookmarks by relaying a message link to the user's DMs."""
 
     def __init__(self, bot: commands.Bot):
```
Comment (anand2312) [inline bot/exts/evergreen/bookmark.py:21]: In #718, all `commands.Bot` type hints were changed to `Bot` from `bot/bot.py`; we should probably use that here (and in the setup function L143) too to stay consistent

**Review by anand2312 — APPROVED** [review]
Tested as well as I could, seems to work well :thumbsup:

File: bot/exts/evergreen/bookmark.py [hunk 8]
```diff
@@ -37,29 +105,42 @@ async def bookmark(
...
+        # Keep track of who has already bookmarked, so users can't spam reactions and cause loads of DMs
+        bookmarked_users = [ctx.author.id]
+        reaction_message = await self.send_reaction_embed(ctx.channel, target_message)
+
+        while True:
+            try:
+                _, user = await self.bot.wait_for("reaction_add", timeout=TIMEOUT, check=event_check)
+            except asyncio.TimeoutError:
+                log.debug("Timed out waiting for a reaction")
+                # No reactions for the last `TIMEOUT` seconds
```
Comment (decorator-factory) [inline bot/exts/evergreen/bookmark.py:134]: This comment is really redundant tbh. 
1. We get a `TimeoutError` while waiting for a reaction to time out
2. There's a logging message explaining what's going on

File: bot/exts/evergreen/bookmark.py [hunk 9]
```diff
@@ -37,29 +105,42 @@ async def bookmark(
...
+                break
+            log.trace(f"{user} has successfully bookmarked from a reaction, attempting to DM them.")
+            await self.action_bookmark(ctx.channel, user, target_message, title)
+            bookmarked_users.append(user.id)
+
+        # Delete the message now that the bot isn't listening to it to save screen space
```
Comment (decorator-factory) [inline bot/exts/evergreen/bookmark.py:140]: The next line makes total sense without the comment. It's how many commands in the bot operate.

File: bot/exts/evergreen/bookmark.py [hunk 10]
```diff
@@ -1,21 +1,89 @@
...
 class Bookmark(commands.Cog):
     """Creates personal bookmarks by relaying a message link to the user's DMs."""
 
-    def __init__(self, bot: commands.Bot):
+    def __init__(self, bot: Bot):
         self.bot = bot
 
+    @staticmethod
+    def build_bookmark_dm(target_message: discord.Message, title: str) -> discord.Embed:
+        """Builds the embed to DM the bookmark requester."""
```
Comment (decorator-factory) [inline bot/exts/evergreen/bookmark.py:27]: the Python Discord style guide tells to phrase the docstrings in an imperative way, i.e.
> Build the embed to DM the bookmark requester.

**Review by decorator-factory — CHANGES_REQUESTED** [review]

**Review by decorator-factory — APPROVED** [review]
L G T M

**Review by jb3 — APPROVED** [review]

**Pull request merged**